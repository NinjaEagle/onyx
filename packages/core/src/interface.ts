import Suite, { isSuite, rootSymbol } from './suite';
import Test from './test';

const root = new Suite('root');
root[rootSymbol] = true;
let currentRoot: Suite = root;

export const getCurrentRoot = () => currentRoot;
export const setCurrentRoot = (v: Suite) => {
  if (!isSuite(v)) { return; }
  currentRoot = v;
};

export { root, currentRoot };

// istanbul ignore next internal
function _it(description: string, fn: () => void, skip = false) {
  if (currentRoot.isRoot()) {
    throw new Error(`"${description}" "it" should not be called outside of "describe" block`);
  }

  const test = new Test(description, fn, skip, currentRoot);

  currentRoot.addChild(test);
  return test;
}

export type ItFn = (description: string, fn: () => void) => Test;
// tslint:disable-next-line:interface-name
export interface It extends ItFn {
  skip: ItFn;
}

const it: It = (description: string, fn: () => void) => _it(description, fn, false);
it.skip = (description: string, fn: () => void) => _it(description, fn, true);

// istanbul ignore next internal
function _describe(description: string, fn: () => void, skip = false): Suite {
  const suite = new Suite(description, skip, currentRoot);

  currentRoot = suite;
  fn();
  if (!suite.skip && suite.children.length <= 0) {
    console.warn(`suite "${suite.getFullDescription()}" doesn't have any child tests or suites.`);
  }
  currentRoot = currentRoot.parent || root;

  currentRoot.addChild(suite);
  return suite;
}

export type DescribeFn = (description: string, fn: () => void) => Suite;
// tslint:disable-next-line:interface-name
export interface Describe extends DescribeFn {
  skip: DescribeFn;
}

const describe: Describe = (description: string, fn: () => void) => _describe(description, fn, false);
describe.skip = (description: string, fn: () => void) => _describe(description, fn, true);

export { it, describe };
