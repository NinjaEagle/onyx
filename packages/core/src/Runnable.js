/**
 * `Runnable` class that abstracts basic functionality shared
 * between the `Test` and `Suite` classes.
 * 
 * @class
 * @param {String} title
 * @param {Function} fn
 */
export default class Runnable {
  title = '';
  fn = null;
  pending = null;
  sync = null;
  async = !sync;

  constructor(title, fn) {
    this.title = title;
    this.fn = fn;
    this.async = fn && fn.length;
    this.sync = !this.async;
    this.pending = false;
  }


  /**
   * Check if this runnable or its parent suite is marked as pending.
   *
   * @private
   */
  _isPending() {
    return this.pending || this.isPending();
  }
}
