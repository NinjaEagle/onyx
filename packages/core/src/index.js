'use strict';

module.exports = core;

function core() {

    const describe = (description, fn) => {
      return new Suite(description, fn);
    };

    // const xdescribe = () => {};

    const it = (description, fn) => {
      return new Test(description, fn);
    };

    // const xit = () => {};
}
