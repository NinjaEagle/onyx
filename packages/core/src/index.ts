module.exports = core;

function core() {
  // Onyx core fn:
}
