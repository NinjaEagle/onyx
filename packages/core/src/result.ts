export enum Status {
  Pending,
  Running,
  Passed,
  Failed,
  Skipped,
}

export default class Result {
  private internalStatus: Status;
  private internalMessages: string[];

  constructor(status?: Status, messages: string | string[] = []) {
    this.internalStatus = status || Status.Pending;
    if (!Array.isArray(messages)) {
      messages = [ messages ];
    }
    this.internalMessages = messages;
  }

  public isDone() {
    return this.internalStatus !== Status.Pending && this.internalStatus !== Status.Running;
  }

  public get status() {
    return this.internalStatus;
  }
  public set status(v: Status) {
    if (this.isDone()) { return; }
    this.internalStatus = v;
  }

  public get messages() {
    return this.internalMessages;
  }
  // No ability to delete messages
  public addMessages(...messages: string[]) {
    if (this.isDone()) { return; }
    this.internalMessages.push(...messages);
  }
}
