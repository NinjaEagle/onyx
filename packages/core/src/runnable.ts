import { EventEmitter } from 'events';
import Result, { Status } from './result';
import Suite from './suite';

export const runnableSymbol = Symbol('isRunnable');

export const isRunnable = (v: unknown): v is Runnable => {
  if (typeof v === 'object' && v === null) { return false; }
  return (v as Runnable)[runnableSymbol];
};

export enum RunnableTypes {
  Runnable = 'runnable',
  Suite = 'suite',
  Test = 'test',
}

/**
 * @class
 * @param {String} description
 * @param {boolean} skip
 */
export default class Runnable extends EventEmitter {
  public description: string;
  public result: Result;
  public skip: boolean;
  public parent: Suite | null;
  public type: RunnableTypes = RunnableTypes.Runnable;
  public [runnableSymbol] = true;

  constructor(description: string, skip?: boolean, parent?: Suite | null) {
    super();
    this.description = description;
    this.result = new Result();
    this.skip = skip || false;
    this.parent = parent || null;
  }

  public doStart(): void {
    this.result.status = Status.Running;
    this.emit('start', this);
  }

  public doEnd(): void {
    this.emit('end', this);
  }

  public doPass(): Result {
    this.result.status = Status.Passed;
    this.emit('pass', this);
    this.doEnd();

    return this.result;
  }

  public doFail(error?: Error | string): Result {
    if (error) {
      this.result.addMessages(String(error));
    }
    this.result.status = Status.Failed;
    this.emit('fail', this, error);
    this.doEnd();

    return this.result;
  }

  public doSkip(): Result { // TODO: add TODO tests support
    this.result.status = Status.Skipped;
    this.emit('skip', this);
    this.doEnd();

    return this.result;
  }

  /**
   * @description Run a `Runnable` instance.
   * @public
   * @return {Result}
   */
  // istanbul ignore next unimplemented
  public run(): Result {
    if (this.skip) {
      return this.doSkip();
    }

    this.doStart();

    return this.doSkip(); // To be replaced with real run function
  }

  /**
   * @description Run `Runnable` instance asynchronously.
   * @public
   * @return {Promise}
   */
  // istanbul ignore next unimplemented
  public async asyncRun(): Promise<Result> {
    if (this.skip) {
      return this.doSkip();
    }

    this.doStart();

    return this.doSkip(); // To be replaced with real run function
  }

  /**
   * @description Check that `Runnable` run has completed and `Result` status is not 'Pending'
   * @public
   * @return {boolean}
   */
  public isDone() {
    return this.result.isDone();
  }

  /**
   * @description return a concatenated description of the current `Runnable` and it's `parent`
   * @public
   * @return {string}
   */
  public getFullDescription(): string {
    if (this.parent && !this.parent.isRoot()) {
      return `${this.parent.getFullDescription()} -> ${this.description}`;
    }
    return this.description;
  }
}
