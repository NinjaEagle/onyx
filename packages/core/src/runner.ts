import { EventEmitter } from 'events';
import Suite from '../src/suite';
import Test from '../src/test';

/**
 * @class
 * @param {Suite} suite
 */
export default class Runner extends EventEmitter {
  public suite: Suite;

  constructor(suite: Suite) {
    super();
    this.suite = suite;
  }

  /**
  * @public
  * @param {Suite} suite
  * @param {Runnable} fn
  */
  public runChildren(suite: Suite, fn: () => any) {}

  /**
   * @public
   * @param {Suite} suite 
   * @param {Runnable} fn 
   */
  public runSuite(suite: Suite, fn: () => any) {
    for (let child of suite.children) {
      if (child.type === 'Suite') {
        this.runSuite((suite), () => {
          fn();
        });
      }
    }

    this.runChildren(suite, () => {
      fn();
    });
  }
}
