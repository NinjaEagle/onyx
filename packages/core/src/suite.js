import Test from './test';

export default class Suite {
  status = '';
  tests = [];

  constructor(status, tests) {
    this.status = status;
    this.tests = tests;
  }

  run() {
    this.status = 'pending';

    this.tests.forEach(test => {
      if(test instanceof Test) {
        test.run();
      }
    });
  }

  asyncRun(...args) {
    return this.run(...args);
  }
};
