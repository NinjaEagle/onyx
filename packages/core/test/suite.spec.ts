import { Status } from '../src/result';
import Runnable from '../src/runnable';
import Suite from '../src/suite';

describe('Suite', () => {
  it('should run children', () => {
    const child = new Suite('child');
    const spy = jest.spyOn(child, 'run');
    const parent = new Suite('parent');
    parent.addChild(child);
    parent.run();

    expect(spy).toHaveBeenCalled();
  });

  it('should push a child to children array', () => {
    const child = new Suite('desc');
    const suite = new Suite('Suite');

    expect(suite.children).toEqual([]);

    suite.addChild(child);
    expect(suite.children.length).toEqual(1);
  });

  // tslint:disable-next-line:max-classes-per-file
  class PassingRunnable extends Runnable {
    public run() {
      this.result.addMessages('OK');
      this.result.status = Status.Passed;
      return this.result;
    }
  }
  it('should pass and collect messages', () => {
    const child = new PassingRunnable('desc');
    const suite = new Suite('Suite');
    suite.addChild(child);
    suite.run();

    expect(suite.result).toMatchSnapshot();
  });

  // tslint:disable-next-line:max-classes-per-file
  class ErroringRunnable extends Runnable {
    public run() {
      this.result.addMessages('FATAL ERROR!');
      this.result.status = Status.Errored;
      return this.result;
    }
  }
  it('should error and collect messages', () => {
    const child = new ErroringRunnable('desc');
    const suite = new Suite('Suite');
    suite.addChild(child);
    suite.run();

    expect(suite.result).toMatchSnapshot();
  });

  // tslint:disable-next-line:max-classes-per-file
  class FailingRunnable extends Runnable {
    public run() {
      this.result.addMessages('FAIL!');
      this.result.status = Status.Failed;
      return this.result;
    }
  }
  it('should fail and collect messages', () => {
    const child = new FailingRunnable('desc');
    const suite = new Suite('Suite');
    suite.addChild(child);
    suite.run();

    expect(suite.result).toMatchSnapshot();
  });
});
