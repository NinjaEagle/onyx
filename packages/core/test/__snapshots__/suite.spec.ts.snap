// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suite get state 1`] = `
Object {
  "done": 0,
  "failed": 0,
  "passed": 0,
  "pending": 1,
  "skipped": 0,
  "total": 1,
}
`;

exports[`Suite get state 2`] = `
Object {
  "done": 1,
  "failed": 0,
  "passed": 1,
  "pending": 0,
  "skipped": 0,
  "total": 1,
}
`;

exports[`Suite should fail and collect messages 1`] = `
Result {
  "internalMessages": Array [
    "FAIL!",
    "undefined",
  ],
  "internalStatus": 3,
}
`;

exports[`Suite should pass and collect messages 1`] = `
Result {
  "internalMessages": Array [
    "OK",
  ],
  "internalStatus": 2,
}
`;
